### PR TITLE
fix(cli): fix hanging tests and harden shim against self-exec loop

### PR DESCRIPTION
**Pull request by fengmk2**
## Problem

Two tests in `vite_global_cli` can hang forever:

- `commands::env::pin::tests::test_check_dev_engines_sync_warns_without_tty`: `check_dev_engines_sync` checks `stdin().is_terminal()` internally, so when the suite runs from an interactive terminal the test blocks on `read_line` waiting for keyboard input.
- `commands::version::tests::detect_system_node_version_returns_version` (intermittent): it spawns `node`, which resolves to the vp shim, while concurrent `#[serial]` tests mutate `PATH`/`VP_HOME` process-wide. With `VP_HOME` pointing elsewhere, `find_system_tool` fails to filter the shim's own bin dir, so the shim resolves itself as the system node and exec's itself in an infinite loop at 100% CPU.

## Fix

- Pass interactivity into `check_dev_engines_sync` as a parameter computed at the call site, so tests exercise the non-TTY and `--force` paths deterministically (new force-path test included).
- Mark `detect_system_node_version_returns_version` as `#[serial]` so no other test mutates the process env while it spawns `node`.
- Harden `find_system_tool`: skip the running executable's directory in the PATH scan, and never return a path that canonicalizes to the running executable. This also protects real users from the self-exec spin (e.g. mis-set `VP_HOME` with `~/.vite-plus/bin` on PATH).

Full suite: 1492 tests pass with no hangs, clippy and rustfmt clean.

## Also includes

- Snap normalizer: strip pnpm's bracketed `[WARN] GET <registry url> ... Will retry` retry lines (the existing rule only matched the TTY thin-space `WARN` form, now written as a visible `\u2009` escape).
- `package.spec.ts`: pin `getNpmRegistry` via partial mock so registry URL assertions pass for developers using a mirror registry.
- Refresh install snaps with pnpm's new "Lockfile passes supply-chain policies" line.

---

> [!NOTE]
> **Medium Risk**
> Shim PATH filtering changes real bypass/system-first behavior (intended fix for infinite exec); pin UX is unchanged at the call site but the devEngines sync path is user-facing when pinning.
> 
> **Overview**
> Fixes **hanging and flaky `vite_global_cli` tests** and **shim self-exec loops** when PATH resolution picks up the vp shim instead of real `node`.
> 
> **`check_dev_engines_sync`** now takes an explicit `interactive` flag (from stdin TTY at the pin call site) instead of calling `stdin().is_terminal()` inside the helper, so tests can force non-TTY and `--force` behavior without blocking on input. New coverage asserts `--force` skips prompts even when `interactive` is true.
> 
> **`find_system_tool`** excludes the running shim’s directory from PATH scanning and refuses a resolved path that canonicalizes to `current_exe()`, closing loops when `VP_HOME` and the shim on PATH disagree (including symlink edge cases).
> 
> **`detect_system_node_version_returns_version`** runs under `#[serial]` so concurrent tests mutating `PATH`/`VP_HOME` do not break the spawned `node` probe.
> 
> Smaller fixes: snap normalizer strips pnpm **`[WARN] GET …`** retry lines; **`package.spec.ts`** mocks `getNpmRegistry` for mirror setups; install snap fixtures include pnpm’s supply-chain lockfile line.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c3ccdcbe1cb5dd890d88fcb1e95a4647db872cc4. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by fengmk2** [comment]
@cursor review

**Comment by fengmk2** [comment]
@codex review

File: crates/vite_global_cli/src/shim/dispatch.rs [hunk 1]
```diff
@@ -1314,7 +1324,20 @@ pub(crate) fn find_system_tool(tool: &str) -> Option<AbsolutePathBuf> {
...
+
+    // Backstop for symlinked shims that evade the directory filters above
+    // (`current_exe` is fully resolved on Linux): never return the running
+    // executable itself.
+    if let Some(self_exe) = &self_exe
+        && let (Ok(resolved_real), Ok(self_real)) =
+            (resolved.as_path().canonicalize(), self_exe.canonicalize())
+        && resolved_real == self_real
+    {
+        return None;
```
[the remark this answers is not in the document]
Comment (fengmk2) [reply crates/vite_global_cli/src/shim/dispatch.rs]: Good catch, fixed in 599f46cf: the backstop now drops the self candidate's directory and retries resolution with the remaining PATH, so a real system tool later in PATH is still found. Added a symlink-based regression test (test_find_system_tool_skips_self_symlink_and_keeps_searching).

**Comment by fengmk2** [comment]
@codex review

File: crates/vite_global_cli/src/shim/dispatch.rs [hunk 2]
```diff
@@ -1293,13 +1293,22 @@ pub(crate) fn find_system_tool(tool: &str) -> Option<AbsolutePathBuf> {
...
         .map(|v| std::env::split_paths(&v).collect())
         .unwrap_or_default();
+
+    // Also skip the directory containing the running shim executable. The
+    // configured bin dir can point elsewhere (e.g. VP_HOME overridden) while
+    // the invoked shim still lives on PATH; without this the lookup resolves
+    // back to the shim itself, which then exec's itself in an infinite loop.
+    let self_exe = std::env::current_exe().ok();
+    if let Some(self_dir) = self_exe.as_deref().and_then(|exe| exe.parent()) {
+        bypass_paths.push(self_dir.to_path_buf());
```
[the remark this answers is not in the document]
Comment (fengmk2) [reply crates/vite_global_cli/src/shim/dispatch.rs]: Fixed in 40b0240d: the upfront exclusion of current_exe's directory is removed entirely; only the canonicalized per-candidate identity check rejects the shim, so a real tool sitting next to vp in a shared bin dir stays discoverable.

File: crates/vite_global_cli/src/shim/dispatch.rs [hunk 3]
```diff
@@ -1310,11 +1319,28 @@ pub(crate) fn find_system_tool(tool: &str) -> Option<AbsolutePathBuf> {
...
+        // Use vite_command::resolve_bin with filtered PATH - stops at first match
+        let search_path = std::env::join_paths(&filtered_paths).ok()?;
+        let resolved = vite_command::resolve_bin(tool, Some(&search_path), &cwd).ok()?;
+        if self_real.is_none() || resolved.as_path().canonicalize().ok() != self_real {
+            return Some(resolved);
+        }
+        let self_candidate_dir = resolved.as_path().parent()?.to_path_buf();
+        let len_before = filtered_paths.len();
+        filtered_paths.retain(|p| *p != self_candidate_dir);
+        if filtered_paths.len() == len_before {
```
[the remark this answers is not in the document]
Comment (fengmk2) [reply crates/vite_global_cli/src/shim/dispatch.rs]: Fixed in 40b0240d. Digging in showed the failure was earlier than the retain comparison: which reports matches from relative PATH entries as relative paths, which AbsolutePathBuf rejects, so resolve_bin errored outright. Relative entries are now resolved against cwd before searching (tilde entries left to which's expansion), and the retain also compares canonicalized dirs. Added test_find_system_tool_skips_self_symlink_in_relative_path_entry.

**Comment by fengmk2** [comment]
@codex review

**Review by wan9chi — APPROVED** [review]
